modules/editor/grammar/post.pegjs

```diff
@@ -0,0 +1,95 @@
+Document
+  = WP_Block_List
+
+WP_Block_List
+  = WP_Block*
+
+WP_Block
+  = WP_Block_Balanced
+  / WP_Block_Html
+
+WP_Block_Balanced
+  = s:WP_Block_Start ts:(!WP_Block_End c:Any { return c })+ e:WP_Block_End
+  { return {
+    blockType: s.blockType,
+    attrs: s.attrs,
+    rawContent: ts.join( '' ),
+  } }
+
+WP_Block_Html
+  = ts:(!WP_Block_Balanced c:Any { return c })+
+  {
+    return {
+      blockType: 'html',
+      attrs: {},
+      rawContent: ts.join('')
+    }
+  }
+
+WP_Block_Start
+  = "<!--" __ "wp:" blockType:WP_Block_Type attrs:WP_Block_Attribute_List _? "-->"
+  { return {
+    type: 'WP_Block_Start',
+    blockType,
+    attrs,
+    text: text()
+  } }
+
+WP_Block_End
+  = "<!--" __ "/wp" __ "-->"
```

Comment:
I see this as necessary for a couple of reasons:

- A fallback so that we can reliably detect more failure conditions.  What if someone edits the HTML(+ comment delimiters) and accidentally transposes part of a block into another block?  We need to be able to throw a parse error here, and if our ending delimiters are just generic `/wp`, we can't really do that.
- WP post content is very much "tag soup" already, and eventually as people start using the block-based editor these comment delimiters will become the same way.  Especially with very long blocks, it needs to be clear both to humans and computers reading the post content not just that a block is ending, but *which* block is ending along with the beginning of a new block.
- If we ever want to be able to support nested blocks this becomes even more of a necessity.  (Conversely, if we never support nested blocks, this doesn't hurt anything.)

As usual I am trying to argue in favor of the most robust approach possible from the very start.  I think that if we don't do this now, we will wish we had, but it'll be too late to change it.